### ISSUE TITLE
Angular formatters vs Angular parsers

### ISSUE THREAD

**Issue by iainbeeston** [Textalk/angular-schema-form-datepicker#6]
## TL;DR

angular-schema-form validates (with tv4) what's visible in the view. angular-schema-form-datepicker converts the dates to ISO8601 format (which json-schema supports) but does not change what is in the view. Because of this, when dates in the form are validated against the json schema (using the view value) validation can fail because it does not conform to the schema.
## PROBLEM

Pickadate uses it's own localisations to convert the date (object) that the user picks to a string, which is displayed in the form input on-screen. For example, right now if I select 16th September 2014, because of my locale I will see "16 September, 2014" in my form.

The date format supported by json-schema is ISO8601 (eg. yyyy-mm-dd), so angular-schema-form-datepicker adds an angularjs formatter function (eg. `ngModel.$formatters.push(fn)`) to datepicker fields, which converts the date selected to a ISO8601 string, and then sets the formatted string on the underlying ngModel. This is the value that developers will see, and usually submit to an API (which presumably defines the json schema used for validation).

So far, so good.

angular-schema-form adds an angularjs $parser function to the ngModel (parser functions are called when an input changes), which calls tv4 and validates that single field. _BUT..._ $parsers get sent the view value, not the model value (e.g. "16 September, 2014" not "2014-09-16"). Because of this angular-schema-form-datepicker is validating what is visible in the view against the json schema, even though this is not what will be submitted to the api. For example, if a pattern is set in the schema for a date-time field, it will fail validation unless the pickadate locale and the json-schema pattern match. Currently this is only an issue if there's no pattern set for the date-time in the schema, but it's conceivable that tv4 might change (in the future) to validate that date-time fields conform to ISO8601.
## SUMMARY

angular-schema-form itself never uses angularjs formatters to convert values - all other form types in angular-schema-form save to the model exactly what is visible in the view. angular-schema-form-datepicker is the only place I have found where the view value is different from the model value, which is why I've raised the problem here.

I believe that angular-schema-form should be validating the value from the model, not the view, because it's the model that will be submitted to the server (and that needs to conform to json-schema). It's also possible to imagine other add-ons in the future which will use formatters.
